File: packages/compiler/src/Parser/tokens.js
import { createToken, Lexer } from "chevrotain";

const WORDS = ["Identifier"];

export default processTokens({
    // note we are placing WhiteSpace first as it is very common thus it will speed up the lexer.
    WhiteSpace: {
        pattern: /\s+/,
        group: Lexer.SKIPPED
    },
    SingleLineComment: {
        pattern: /\/\/.*?\n/,
        group: Lexer.SKIPPED
    },
    Semicolon: /;/,
    Assignment: /=/, // longer_alt: AssignmentOperator
    LParen: /\(/,
    RParen: /\)/,
    // "keywords" appear before the Identifier
    MutabilityModifier: {
        pattern: /const|mutable/,
        longer_alt: "Identifier"
    },
    ExportModifier: {
        pattern: /export/,
        longer_alt: "Identifier"
    },
    Identifier: /[a-zA-Z_][0-9a-zA-Z_]*/,
    AdditionOperator: /[+\-]/,
    ProductOperator: /[*\/]/,
    ExponentialOperator: /\^/,
    UnaryOperator: /[!\-]/,
    Number: /-?(?:[0-9]|[1-9][0-9]+)(?:\.[0-9]+)?(?:[eE][-+]?[0-9]+)?\b/,
    String: /"(?:[^"\\]|.)*"/
});

function processTokens(tokenObj) {
    function tok(name, value) {
        return createToken(Object.assign({ name },
            value instanceof RegExp ? { pattern: value } : value));
    }

    // Deal with words first to replace longer_alt fields
    const wordsObj = {};
    for (const w of WORDS) {
        // TODO: Error if doesn't exist
        wordsObj[w] = tok(w, tokenObj[w]);
    }

    for (const key of Object.keys(tokenObj)) {
        if (key in wordsObj) {
            tokenObj[key] = wordsObj[key]
        } else {
            const value = tokenObj[key];
            if (value.longer_alt) { // TODO: Error if doesn't exist
                value.longer_alt = wordsObj[value.longer_alt];
            }
            tokenObj[key] = tok(key, value);
        }
    }
    return tokenObj;
}

File: packages/compiler/src/Parser/Tokens.fs
module Ikigai.Compiler.Parser.Tokens

open System.Text.RegularExpressions
open Fable.Core
open Fable.Core.JsInterop
open Chevrotain

[<Import("*", "chevrotain")>]
let chevrotain: IExports = jsNative

// Fable adds g flag by default, which is not supported by chevrotain
[<Emit("new RegExp($0)")>]
let inline regex pattern: U2<string,Regex> = jsNative

type Token =
    [<Import("createToken", "chevrotain")>]
    [<Emit("$0({ name: $1, pattern: $2, {{group: $3,}} {{longer_alt: $4,}} })")>]
    static member Create(name: string, pattern: U2<string,Regex>, ?group: string, ?longer_alt: TokenType): TokenType = jsNative

let WhiteSpace = Token.Create("WhiteSpace", regex "s+", chevrotain.Lexer.SKIPPED)
let SingleLineComment = Token.Create("SingleLineComment", regex (@"\/\/.*?\n"), chevrotain.Lexer.SKIPPED)
let Semicolon = Token.Create("Semicolon", !^";")
let Assignment = Token.Create("Assignment", !^"=")
let LParen = Token.Create("LParen", !^"(")
let RParen = Token.Create("RParen", !^")")

// Keywords
let Identifier = Token.Create("Identifier", regex "[a-zA-Z_][0-9a-zA-Z_]*")
let MutabilityModifier = Token.Create("MutabilityModifier", regex "const|mutable", longer_alt=Identifier)
let ExportModifier = Token.Create("ExportModifier", regex "export", longer_alt=Identifier)

// Operators
let AdditionOperator = Token.Create("AdditionOperator", regex (@"[+\-]"))
// ProductOperator: /[*\/]/,
// ExponentialOperator: /\^/,
// UnaryOperator: /[!\-]/,

// Literals
let Number = Token.Create("Number", regex @"-?(?:[0-9]|[1-9][0-9]+)(?:\.[0-9]+)?(?:[eE][-+]?[0-9]+)?\b")
let String = Token.Create("String", regex @"""(?:[^""\\]|.)*""")

let TOKENS =
    [|
    WhiteSpace
    SingleLineComment
    Semicolon
    Assignment
    LParen
    RParen
    MutabilityModifier
    ExportModifier
    Identifier
    AdditionOperator
    Number
    String
    |]